### PR TITLE
Support shared domain

### PR DESCRIPTION
**Pull request by userzj**
https://lore.kernel.org/linux-erofs/20220902105305.79687-1-zhujia.zj@bytedance.com/T/#t
Signed-off-by: Jia Zhu <zhujia.zj@bytedance.com>

### CODE REVIEW

**Comment by hsiangkao** [comment]
``
golangci-lint run
pkg/utils/erofs/erofs_linux.go:21: File is not `gofmt`-ed with `-s` (gofmt)
		if err := mount("erofs", mountPoint, "erofs", 0, "fsid=" + fsID); err != nil {
``

File: config/daemonconfig.go [hunk 1]
```diff
@@ -161,7 +161,7 @@ func NewDaemonConfig(fsDriver string, cfg DaemonConfig, imageID, snapshotID stri
 			backendConfig = &cfg.Config.BackendConfig
 			fscacheID := erofs.FscacheID(snapshotID)
 			cfg.ID = fscacheID
-			cfg.DomainID = fscacheID
+			cfg.DomainID = "global-domain"
```
Comment (changweige) [inline config/daemonconfig.go:164]: Is domain ID specific to kernel/erofs? In other words, is it a predefined keyword in the Kernel?

File: pkg/utils/erofs/erofs_linux.go [hunk 2]
```diff
@@ -15,10 +15,12 @@ import (
 func Mount(bootstrapPath, fsID, mountPoint string) error {
 	mount := unix.Mount
 
-	opts := "fsid=" + fsID
+	opts := fmt.Sprintf("domain_id=%s,fsid=%s", "global-domain", fsID)
 	logrus.Infof("Mount erofs to %s with options %s", mountPoint, opts)
 	if err := mount("erofs", mountPoint, "erofs", 0, opts); err != nil {
-		return errors.Wrapf(err, "failed to mount erofs")
+		if err := mount("erofs", mountPoint, "erofs", 0, "fsid="+fsID); err != nil {
+			return errors.Wrapf(err, "failed to mount erofs")
```
Comment (changweige) [inline pkg/utils/erofs/erofs_linux.go:22]: Why do you retry another mount if the L20 mount operation fails? It's better to add a comment to tell necessity and intention.

File: pkg/utils/erofs/erofs_linux.go [hunk 3]
```diff
@@ -15,10 +15,12 @@ import (
 func Mount(bootstrapPath, fsID, mountPoint string) error {
 	mount := unix.Mount
 
-	opts := "fsid=" + fsID
+	opts := fmt.Sprintf("domain_id=%s,fsid=%s", "global-domain", fsID)
```
Comment (changweige) [inline pkg/utils/erofs/erofs_linux.go:18]: It's better to make "global-domain" as a global constant

File: config/daemonconfig.go (hunk 1 above)
Comment (userzj) [reply config/daemonconfig.go:164]: Yup, @domain_id is a mount option of EROFS.

File: pkg/utils/erofs/erofs_linux.go (hunk 3 above)
Comment (userzj) [reply pkg/utils/erofs/erofs_linux.go:18]: Thanks, I'll revise it.

File: pkg/utils/erofs/erofs_linux.go (hunk 2 above)
Comment (userzj) [reply pkg/utils/erofs/erofs_linux.go:22]: If a mount with the option "domain_id" fails, this indicates that the kernel may not support EROFS shared domain, and will fall back to normal mount. I'll add a comment about that. Thanks.

File: config/daemonconfig.go (hunk 1 above)
Comment (changweige) [reply config/daemonconfig.go:164]: Actually, I mean "global-domain" :-)

File: pkg/utils/erofs/erofs_linux.go (hunk 2 above)
Comment (changweige) [reply pkg/utils/erofs/erofs_linux.go:22]: I see. Can it check return error code like unsupported or invalid ?

File: config/daemonconfig.go (hunk 1 above)
Comment (userzj) [reply config/daemonconfig.go]: 'global-domain' is not a keyword in the kernel. This is just a name(id) specified by us. In future work, 'domain_id' will be specified by the user.

File: pkg/utils/erofs/erofs_linux.go [hunk 4]
```diff
@@ -12,15 +12,24 @@ import (
 	"golang.org/x/sys/unix"
 )
 
+const globalDomainName = "global-domain"
```
Comment (changweige) [inline pkg/utils/erofs/erofs_linux.go:15]: Um. erofs package should import the constant from daemonconfig.go

File: pkg/utils/erofs/erofs_linux.go [hunk 5]
```diff
@@ -12,15 +12,24 @@ import (
...
+
 func Mount(bootstrapPath, fsID, mountPoint string) error {
 	mount := unix.Mount
+	var err error
 
-	opts := "fsid=" + fsID
+	opts := fmt.Sprintf("domain_id=%s,fsid=%s", globalDomainName, fsID)
 	logrus.Infof("Mount erofs to %s with options %s", mountPoint, opts)
-	if err := mount("erofs", mountPoint, "erofs", 0, opts); err != nil {
+	err = mount("erofs", mountPoint, "erofs", 0, opts)
```
Comment (changweige) [inline pkg/utils/erofs/erofs_linux.go:23]: We still can improve the code like:
```
if err := mount("erofs", mountPoint, "erofs", 0, opts); err!=nil {
	if errors.Is(err, unix.EINVAL) {
		err = mount("erofs", mountPoint, "erofs", 0, "fsid="+fsID)
                 if err != nil {
                      return errr
}
	}

         return errors.Wrapf(err, "failed to mount erofs")
}

```

Comment (userzj) [reply pkg/utils/erofs/erofs_linux.go]: Thanks.

**Review by changweige — APPROVED** [review]
LGTM, thanks

**Pull request merged**